File: mobile_ext/flet_open_file/flet_open_file/open_file.py
"""Python side of the flet_open_file extension — a Flet Service that opens a
local file with the OS default handler.

Used by the in-app updater: opening a .apk this way makes Android hand it
straight to the package installer, skipping the share-sheet app chooser that
ft.Share produced. Modeled on flet_local_auth / flet_secure_storage (same Flet
0.85 @control + Service + _invoke_method pattern).

NOTE: the system "Do you want to install this update?" dialog and the one-time
"Allow from this source" grant are Android security guarantees and CANNOT be
bypassed — this only removes the chooser step in front of them.
"""
from flet.controls.base_control import control
from flet.controls.services.service import Service


@control("OpenFile")
class OpenFile(Service):
    """Opens a local file path with the platform's default handler."""

    async def open(self, path: str) -> str:
        """Open `path`. Returns the plugin's result type as a string
        ("done", "noAppToOpen", "permissionDenied", "fileNotFound", "error").
        Never raises — failures come back as a string so the caller can fall
        back to the share sheet."""
        return str(await self._invoke_method("open", {"path": path}) or "")

File: mobile_ext/flet_open_file/flutter/flet_open_file/lib/src/open_file.dart
import 'dart:async';

import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:open_filex/open_filex.dart';

/// Dart side of flet_open_file. Mirrors flet_local_auth's service pattern.
///
/// Opening a downloaded .apk here makes Android resolve it to the package
/// installer directly, instead of the share-sheet chooser ft.Share produced.
/// Requires REQUEST_INSTALL_PACKAGES in the manifest (see build-apk.yml); the
/// system install confirmation and the one-time "unknown sources" grant still
/// apply and cannot be bypassed by any app.
class OpenFileService extends FletService {
  OpenFileService({required super.control});

  @override
  void init() {
    super.init();
    debugPrint("OpenFileService(${control.id}).init");
    control.addInvokeMethodListener(_invokeMethod);
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    switch (name) {
      case "open":
        // Never let an exception cross the bridge — the caller falls back to
        // the share sheet on anything other than "done".
        try {
          final path = (args?["path"] ?? "") as String;
          if (path.isEmpty) return "error";
          final res = await OpenFilex.open(path);
          return res.type.name;
        } catch (e) {
          debugPrint("OpenFileService.open error: $e");
          return "error";
        }
      default:
        throw Exception("Unknown OpenFile method: $name");
    }
  }

  @override
  void dispose() {
    control.removeInvokeMethodListener(_invokeMethod);
    super.dispose();
  }
}
